content/posts/observability-in-the-cloud.md

```diff
@@ -0,0 +1,90 @@
+We iteratively go back to the logs, to grab more pieces of evidence and approximate the state of the system when the error occurred.
+Since we can use debuggers to investigate the problems, and the possible problem space is usually somewhat limited, we can rely on this telemetry, which is usually highly aggregated.
+
+However, this aggregated telemetry does not carry enough information when troubleshooting applications designed for the clouds or refactored from monolithic applications into a microservice architecture.
+
+In a microservice architecture, we have a lot more different runtimes handling work and interacting via the network.
+Additionally, the following practices are being adopted more widely, when we design applications for clouds:
+
+- dynamic scaling and disposable application instances
+- asynchronous processing, by using queues and streaming services
+- blue/green, canary or rolling deployments
+- feature flags
+- shorter deployment cycles
+
+These changes lead to a more complex application landscape, which makes debugging increasingly difficult.
+Instead of running one service in a debugger and recreating its state, we may need to start various services and try to reproduce state in each of them.
+Our telemetry may also become less expressive and we may have to ask ourselves additional questions, such as:
+
+- has a deployment just happened?
+- what versions are actually deployed?
+- which versions (if multiple are running) show anomalous behaviour
+- which feature flags were enabled for the requests/tasks that failed?
+
+To ensure, we still get some meaningful insights from our telemetry, we need to annotate it with additional information such as build ID, hostname, IP address, enabled feature flags and more.
+This data enrichment brings about new problems: adding more dimensions to our telemetry data makes the storage and querying systems much slower and/or more expensive.
+At the same time, we still need to make a number of queries to correlate information from multiple source systems (microservices) to slowly piece together enough context to guess what happened.
+
+## Closing the Observability Gap in Distributed Cloud Systems
+
+Unfortunately, this gap cannot be closed by the simple addition of a new tool or library.
+Instead, there are two primary challenges that should be tackled by site reliability engineers who aim to improve observability in their systems.
+
+First, each individual service should be instrumented to keep track of its context. Furthermore, each service should emit arbitrarily wide events containing the relevant context when a request/task processing finishes (successfully or with an exception).
+To do this, whenever a workload is started (e.g. a request enters a system), an event context object should be initialized and prepopulated with all relevant data.
```

Comment:
```suggestion
To do this, whenever a workload is processed (e.g. a request enters a system), an event context object should be initialized and prepopulated with all relevant data.
```

The start of a workload, in my intuition, would rather refer to the startup of an application instead of it processing a request.